python-checks/src/test/resources/checks/membershipTestSupportCheck.py

```diff
@@ -0,0 +1,225 @@
+from some_module import unknown
+from typing import Sequence, Iterable, Container, Callable, overload
+
+def produces_sequence() -> Sequence:
+    ...
+
+def produces_iterable() -> Iterable:
+    ...
+
+def produces_container() -> Container:
+    ...
+
+class Empty:
+    pass
+def produces_empty() -> Empty:
+    ...
+
+42 in 1337 # Noncompliant {{Change the type of 1337; type int does not support membership protocol.}}
+#  ^^ ^^^^< {{The result value of this expression does not support the membership protocol.}}
+
+42 not in 1337 # Noncompliant
+#  ^^^^^^ ^^^^< {{The result value of this expression does not support the membership protocol.}}
+
+if 42 in 1337: # Noncompliant
+#     ^^ ^^^^<
+    pass
+
+42 in int(this_is() + a_long() + # Noncompliant {{Change the type for the target expression of `in`; type int does not support membership protocol.}}
+#  ^^ ^[el=+2;ec=29]<
+      multiline_expression())
+
+def g():
+    list_or_int = []
+    if unknown():
+        list_or_int = 42
+
+    unknown() in list_or_int # Compliant: At least one of the members of this union type supports the membership protocol
+
+    bool_or_int = True
+    if unknown():
+        bool_or_int = 42
+    # None of the types in this union type support the membership protocol
+    unknown() in bool_or_int # Noncompliant
+
+def f():
+    x = 1337
+    42 in x # Noncompliant
+    y = 42
+    y in 1337 # Noncompliant
+    x in y # Noncompliant
+
+    x in unknown() # Compliant
+    unknown() in x # Noncompliant
+    unknown() in unknown() # Compliant
+
+    x in [] # Compliant
+    x in () # Compliant
+    x in {} # Compliant
+    x in {v for v in "abc"} # Compliant
+    x in frozenset()
+    x in dict() # Compliant
+    x in {k: 42 for k in "abc"} # Compliant
+    x in produces_container() # Compliant
+    x in produces_iterable() # Compliant
+    x in produces_sequence() # Compliant
+    # TODO: Be more strict regarding declared types?
```

Comment:
For declared types, we have a rule called `ConfusingTypeChecking` which keeps the logic from the type-checking BUG rules but only watches for declared types.
Note that `ConfusingTypeChecking` raises a "Code smell" instead of a "bug".

So I would suggest editing the TODO comment with: Add this to `ConfusingTypeChecking` rule. Also, I think you can create a ticket in the backlog for another sprint.

Reply:
I updated the comment and created a ticket here: https://sonarsource.atlassian.net/browse/SONARPY-1420?atlOrigin=eyJpIjoiMjI2MmNlMDYzY2JmNGFlOThkYzNmZTZmMTM1OGFmZTEiLCJwIjoiaiJ9